File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution { 
public:
    void Dfs(TreeNode* root, int& k, int& answer) {
        if (!root) {
            return;
        }
        Dfs(root->left, k, answer);
        k--;
        if (k == 0) {
            answer = root->val;
            return;
        }
        Dfs(root->right, k, answer);
    }
    int kthSmallest(TreeNode* root, int k) {
        int answer;
        Dfs(root, k, answer);
        return answer;
    }
};

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
#     def __init__(self):
#         # Initialize instance variables
#         self.res = None
#         self.k = 0
    
#     def dfs(self, root):
#         if root is None:
#             return
#         self.dfs(root.left)
#         self.k -= 1
#         if self.k == 0:
#             self.res = root.val
#             return
#         self.dfs(root.right)
        
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        # self.res = None
        # self.k = k
        # self.dfs(root)
        # return self.res
        stack = []
        curr = root
        while curr or stack:
            while curr:
                stack.append(curr)
                curr = curr.left
            curr = stack.pop()
            k -= 1
            if k == 0:
                return curr.val
            curr = curr.right